hypershift-operator/controllers/machinedeployment/controller.go

```diff
@@ -0,0 +1,216 @@
+		r.recorder.Eventf(machineDeployment, corev1.EventTypeWarning, "ReconcileError", "%v", err)
+		// we don't return here so we want to attempt to patch the machine regardless of an error.
+	}
+
+	if err := r.Client.Patch(ctx, machineDeployment, originalMachineDeploymentToPatch); err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to patch machineDeployment: %v", err)
+	}
+
+	return result, err
+}
+
+func (r *Reconciler) reconcile(ctx context.Context, machineDeployment *clusterv1.MachineDeployment) (ctrl.Result, error) {
+	log := ctrl.LoggerFrom(ctx)
+	log.V(3).Info("Reconciling MachineDeployment", "machineDeployment", machineDeployment.Name)
+
+	// Resolve AWSMachineTemplate first
+	awsMachineTemplate, err := utils.ResolveAWSMachineTemplate(ctx, r.Client, machineDeployment)
+	if err != nil {
+		log.V(4).Info("Skipping MachineDeployment", "namespace", machineDeployment.Namespace, "name", machineDeployment.Name, "reason", err.Error())
+		// Not an error - this MachineDeployment may not be for AWS platform
+		return ctrl.Result{}, nil
+	}
+
+	// Check if AWSMachineTemplate.Status.Capacity is already populated by CAPA
+	if len(awsMachineTemplate.Status.Capacity) > 0 {
+		log.V(3).Info("AWSMachineTemplate already has Status.Capacity set by CAPA, skipping annotation workaround",
+			"namespace", awsMachineTemplate.Namespace, "name", awsMachineTemplate.Name)
+		return ctrl.Result{}, nil
+	}
+
+	// CAPA doesn't support scale-from-zero yet, use workaround
+	log.V(3).Info("AWSMachineTemplate has no Status.Capacity, applying annotation workaround",
+		"namespace", awsMachineTemplate.Namespace, "name", awsMachineTemplate.Name)
+
+	// Extract instance type
+	instanceType, err := utils.ExtractInstanceType(awsMachineTemplate)
+	if err != nil {
+		log.Error(err, "Failed to extract instance type")
+		r.recorder.Eventf(machineDeployment, corev1.EventTypeWarning, "FailedUpdate", "Failed to extract instance type: %v", err)
+		return ctrl.Result{}, err
+	}
+
+	// Resolve AWS region
+	region, err := utils.ResolveRegion(ctx, r.Client, machineDeployment)
+	if err != nil {
+		log.Error(err, "Failed to resolve AWS region")
+		r.recorder.Eventf(machineDeployment, corev1.EventTypeWarning, "FailedUpdate", "Failed to resolve AWS region: %v", err)
+		return ctrl.Result{}, err
+	}
+
+	// Create AWS client with region validation
+	awsClient, err := awsclient.NewValidatedClient(ctx, region, r.CredentialsFile, r.RegionCache)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("error creating aws client: %w", err)
+	}
+
+	// Get instance type information
+	instanceTypeInfo, err := r.InstanceTypesCache.GetInstanceType(ctx, awsClient, region, instanceType)
+	if err != nil {
+		log.Error(err, "Unable to set scale from zero annotations: unknown instance type", "instanceType", instanceType)
+		log.Error(nil, "Autoscaling from zero will not work. To fix this, manually populate machine annotations for your instance type", "annotations", []string{cpuKey, memoryKey, gpuKey})
```

Comment:
It seems this information might be more useful/visible in the event, rather than in the log.